### PR TITLE
parser, ddl: support decoding binary literal in set/enum

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/31338

Problem Summary:

https://github.com/pingcap/tidb/issues/31338#issuecomment-1033414941:

> This is because the binary literal like 0xD2BB in enum/set can only be decoded to utf8mb4. When the result is written back to the client, an error occurs because 0xD2BB cannot be encoded with gbk. Thus, the result is truncated.

### What is changed and how it works?

After this PR, the enum/set values are saved by parser along with a flag, which is used to indicate that whether it is a binary literal:

```
// FieldType records field type information.
type FieldType struct {
        // ...
	// elems is the element list for enum and set type.
	elems            []string
	elemsIsBinaryLit []bool
}
```

In the stage of `buildTableInfo`, we decode column's `FieldType.elems` according to the column charset to UTF-8 values, and save it. Thus, the enum/set values are always valid UTF-8, nothing is truncated.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the incorrect result of `SHOW CREATE TABLE` when specifying binary literals as enum/set values.
```

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -820,6 +821,23 @@ func setCharsetCollationFlenDecimal(tp *types.FieldType, colName, colCharset, co
...
+	enc := charset.FindEncoding(chs)
+	for i, elem := range tp.GetElems() {
+		if !tp.GetElemIsBinaryLit(i) {
+			continue
+		}
+		s, err := enc.Transform(nil, hack.Slice(elem), charset.OpDecodeReplace)
+		if err != nil {
+			logutil.BgLogger().Warn("decode enum binary literal to utf-8 failed", zap.Error(err))
+		}
+		tp.SetElem(i, string(hack.String(s)))
```
Comment (Defined2014) [inline ddl/ddl_api.go:837]: maybe `string(s)` is enough

Comment (tangenta) [reply ddl/ddl_api.go:837]: Then it will copy another string...

File: cmd/explaintest/r/new_character_set_invalid.result [hunk 2]
```diff
@@ -21,3 +21,33 @@ a	b	c
 ?	?	?
 中文?中文	asdf?fdsa	字符集?字符集
 @@	@@	@@
+drop table t;
+create table t(f set(0xD2BB, 0xC8FD), e enum(0xBAEC,0x6A59));
+show create table t;
+Table	Create Table
+t	CREATE TABLE `t` (
+  `f` set('һ','??') DEFAULT NULL,
+  `e` enum('??','jY') DEFAULT NULL
```
Comment (Defined2014) [inline cmd/explaintest/r/new_character_set_invalid.result:30]: MySQL only have one '?', do we need to compatible with it?

File: cmd/explaintest/t/new_character_set_invalid.test [hunk 3]
```diff
@@ -15,3 +15,16 @@ insert into t values ('À', 'ø', '😂');
 insert into t values ('中文À中文', 'asdføfdsa', '字符集😂字符集');
 insert into t values (0x4040ffff, 0x4040ffff, 0x4040ffff);
 select * from t;
+
```
Comment (Defined2014) [inline cmd/explaintest/t/new_character_set_invalid.test:18]: I think `set @@sql_mode = default` is better

File: parser/types/field_type.go [hunk 4]
```diff
@@ -53,7 +53,8 @@ type FieldType struct {
 	// collate represent collate rules of the charset
 	collate string
 	// elems is the element list for enum and set type.
-	elems []string
+	elems            []string
+	elemsIsBinaryLit []bool
```
Comment (xhebox) [inline parser/types/field_type.go:57]: I should have documented that `jsonFieldType` is ought to be modified together with `FieldType` to make (de)serialization work after that refactor PR.

Maybe you can do that in this PR?

File: parser/parser.y [hunk 5]
```diff
@@ -11916,11 +11917,12 @@ StringType:
...
-		fieldLen := len(tp.GetElems()) - 1 // set_flen = sum(ele_flen) + number_of_ele - 1
-		for i := range tp.GetElems() {
-			tp.SetElem(i, strings.TrimRight(tp.GetElem(i), " "))
-			fieldLen += len(tp.GetElem(i))
+		tp.SetElems(make([]string, len(elems)))
+		fieldLen := len(elems) - 1 // set_flen = sum(ele_flen) + number_of_ele - 1
+		for i, e := range elems {
+			trimmed := strings.TrimRight(e.Value, " ")
+			tp.SetElemWithIsBinaryLit(i, trimmed, e.IsBinaryLiteral)
+			fieldLen += len(trimmed)
```
Comment (xhebox) [inline parser/parser.y:11925]: Why it is not `len(tp.GetElem())`?

Comment (tangenta) [reply parser/parser.y:11925]: They are the same string.

File: cmd/explaintest/r/new_character_set_invalid.result (hunk 2 above)
Comment (tangenta) [reply cmd/explaintest/r/new_character_set_invalid.result:30]: MariaDB also has 2 question marks... Anyway, let's do it when it is necessary.

File: cmd/explaintest/t/new_character_set_invalid.test (hunk 3 above)
Comment (tangenta) [reply cmd/explaintest/t/new_character_set_invalid.test:18]: Done.

File: parser/types/field_type.go (hunk 4 above)
Comment (tangenta) [reply parser/types/field_type.go:57]: OK, thanks for pointing out.

File: parser/parser.y (hunk 5 above)
Comment (xhebox) [reply parser/parser.y:11925]: I knew. My point is that `tp.GetElem()` makes less assumption about what `SetElemWithIsBinaryLit` is doing. I won't insist on the change.

**Review by Defined2014 — APPROVED** [review]

File: parser/parser.y (hunk 5 above)
Comment (tangenta) [reply parser/parser.y:11925]: Actually the `fieldLen` should not be calculated here because it may change after decoding. Let's remove it.

**Review by xhebox — APPROVED** [review]

**Comment by Defined2014** [comment]
@tangenta Please update mysql-test

**Pull request merged**